### PR TITLE
[INTEL MKL] Execute small gemm's single threaded.

### PR DESCRIPTION
**Pull request by Srini511**
This PR introduces a simple cost metric to determine if dnnl_sgemm needs to be run with 1 thread or the whole threadpool.

Please note: Please dont merge this PR until https://github.com/tensorflow/tensorflow/pull/47543 is merged

### CODE REVIEW

**Comment by Srini511** [comment]
@penpornk I tried to fix the merge conflict online and it has cancelled the cla check. Can you please help resolve this?

**Comment by Srini511** [comment]
@googlebot I fixed it

**Comment by penpornk** [comment]
Manually changing CLA to yes because the two commits in this PR are from the same github account which has CLA.

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 1]
```diff
@@ -33,8 +33,17 @@ using mkldnn::stream;
 
 namespace tensorflow {
 
+#define L1_SIZE 32 * 1024
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:36]: Just FYI, TensorFlow can read cache sizes through Eigen, although that may come with some latency costs.
https://github.com/tensorflow/tensorflow/blob/add0a785e5af7d39cbb6801e2aefd83db39738b0/tensorflow/core/kernels/conv_grad_ops_3d.cc#L395

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 2]
```diff
@@ -33,8 +33,17 @@ using mkldnn::stream;
...
 typedef Eigen::ThreadPoolDevice CPUDevice;
 
+inline bool ExecuteSingleThreadedGemm(int m, int n, int k) {
+  // Ideally we would like to determine blocking and then come up with
+  // a heuristic but what we are targetting are very small models whose
+  // total size is < few L1's. So we will do this simple calculation
+  // to determine if the MM should be run on a single thread.
+  return ((sizeof(float) * (m * n + k * (m + n))) < L1_SIZE * 8);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:44]: Just curious, why is `L1_SiZE` multiplied by 8? If it's some magic constant, please say so in the comment (in the code).

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and I'm sorry for the delay!

**Review by penpornk — APPROVED** [review]
@Srini511 Since the branch cut is in two days, I'll just pull the PR in and make the modifications myself to save time. I'd appreciate if you could help answer the question about `L1_SIZE * 8` later when you have time though. :)

**Comment by Srini511** [comment]
@penpornk , Yes 8 is just a magic number that worked with internal workloads. I had made the changes internally and was testing them on specific models. Thanks!

**Pull request merged**